### PR TITLE
miri engine: Fix run-time validation

### PR DESCRIPTION
**Pull request by RalfJung**
This fixes all false positives that came up when actually enabling this in miri.

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/interpret/validity.rs [hunk 1]
```diff
@@ -554,12 +560,16 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
...
-                    // The closure is not local, so we cannot get the name
-                    PathElem::ClosureVar(Symbol::intern(&field.to_string()))
+                    if let Some(freevar) = self.tcx.with_freevars(
+                        node_id,
+                        |fv| fv.get(field).map(|field| *field))
+                    {
+                        return PathElem::ClosureVar(self.tcx.hir.name(freevar.var_id()));
+                    }
                 }
+                // The closure is not local, or the freevars don't match up (seen for a generator!),
```
Comment (oli-obk) [inline src/librustc_mir/interpret/validity.rs]: `tcx.mir_optimized(def_id).upvar_decls[upvar_index].debug_name` should do it?

Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: There is [no such method](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/?search=mir_optimized)?

Comment (oli-obk) [reply src/librustc_mir/interpret/validity.rs]: sorry, was on mobile

https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.optimized_mir

Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: What will that do if we donÄt have libstd MIR?

Comment (oli-obk) [reply src/librustc_mir/interpret/validity.rs]: uhm... I don't see how that can happen. How would you even see a libstd closure without its MIR being available?

Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: Fair enough.

However, for a generator, I am still seeing

> index out of bounds: the len is 0 but the index is 0

So I'll have to keep some kind of fallback.

Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: Done, and even added a test for it. ;)

File: src/librustc_mir/interpret/validity.rs [hunk 2]
```diff
@@ -510,9 +511,15 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
                         // This is the size in bytes of the whole array.
                         let size = Size::from_bytes(ty_size * len);
 
-                        match self.memory.read_bytes(dest.ptr, size) {
+                        // In run-time mode, we accept points in here.  This is actually more
```
Comment (oli-obk) [inline src/librustc_mir/interpret/validity.rs]: points -> pointers

File: src/librustc_mir/interpret/memory.rs [hunk 3]
```diff
@@ -842,6 +842,28 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> Memory<'a, 'mir, 'tcx, M> {
...
+        allow_ptr: bool,
+    ) -> EvalResult<'tcx> {
+        // Empty accesses don't need to be valid pointers, but they should still be non-NULL
+        let align = Align::from_bytes(1, 1).unwrap();
+        if size.bytes() == 0 {
+            self.check_align(ptr, align)?;
+            return Ok(());
+        }
+        let ptr = ptr.to_ptr()?;
+        self.get_bytes_with_undef_and_ptr(ptr, size, align)?;
```
Comment (oli-obk) [inline src/librustc_mir/interpret/memory.rs]: what's the point of this call? bounds and align checking? A comment would be nice

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by RalfJung** [comment]
Fixed nits.

**Review by oli-obk — APPROVED** [review]

**Comment by RalfJung** [comment]
Oh dang, miri's test suite fails with full MIR. Investigating.

**Comment by RalfJung** [comment]
This should fix it.

@bors r=oli-obk

**Comment by RalfJung** [comment]
Fixed conflicts.

@bors r=oli-obk

**Pull request merged**